lib/mailbox.mli

```diff
@@ -59,40 +187,172 @@ module Domain : sig
+  (** Same as {!literal} but raises an [Invalid_argument] instead to return [None]. *)
 
   val domain : 'a domain t
+  (** Kind of domain. *)
+
   val ipv4 : Ipaddr.V4.t t
+  (** Kind of {!Literal_domain.ipv4}. *)
+
   val ipv6 : Ipaddr.V6.t t
+  (** Kind of {!Literal_domain.ipv6}. *)
+
   val extension : (string * string) t
-  val default : literal t
+  (** Kind of {!Literal_domain.extension}. *)
+
+  val default : string t
+  (** Kind of {!literal}. *)
 
   val make : 'a t -> 'a -> Rfc5322.domain option
-  val make_exn : 'a t -> 'a -> Rfc5322.domain
+  (** [make kind v] returns a safe domain. It can fail if an user-defined
+     literal-domain ({!Literal_domain.extension}), a {!literal} domain or a
+     {!domain} don't follow standards:
+
+     {ul
+     {- for a {!Literal_domain.extension}, [make] returns [None] if
+     {!Literal_domain.make} returns [None]}
+     {- for a {!literal}, [make] returns [None] if {!literal} returns [None]}
+     {- for a {!domain}, [make] returns [None] if list of {!atom} is empty}} *)
+
+  val v : 'a t -> 'a -> Rfc5322.domain
+  (** Same as {!make} but raises an [Invalid_argument] instead [None]. *)
+
+  val to_string : Rfc5322.domain -> string
+  (** [to_string x] returns a string which represents [x] as is it in a e-mails. *)
 end
 
 module Local : sig
+  (** Local part of a mailbox is a non-empty list of {!word} elements.
+      You can construct local-part like this:
+
+      {[
+        let local = Local.(v [ w "romain"; w "calascibetta" ]) ;;
+        val local : local = [ `Atom "romain"; `Atom "calascibetta" ]
+      ]}
+
+      Spaces and control characters are allowed in the local-part:
+
+      {[
+        let local = Local.(v [ w "Romain Calascibetta"; w "to+mrmime" ]) ;;
+        val local : local = [ `String "Romain Calascibetta"; `Atom "to+mrmime" ]
+        Local.to_string local ;;
+        - : string = "\"Romain Calascibetta\".to+mrmime"
+      ]}
+
+      NOTE: [+] permits your MUA to tag e-mails received with this local-part.
+
+      Valid UTF-8 string is allowed according RFC 6532 - and will be
+      surrounded by double-quote. *)
+
   type 'a local =
     | [] : Peano.z local
     | ( :: ) : word * 'a local -> 'a Peano.s local
 
   val w : string -> word
+  (** {!w} produces a {b safe} word. {!w} will try to escape control characters
+     and verify if contents respects standards. Otherwise, {!w} raises an
+     [Invalid_argument]. [`Word] produced by {!w} can be surrounded by
+     double-quote. *)
+
   val word : string -> word option
+  (** [word x] tries to normalize [x] as a [`Word] according RFC 5322. It
+     returns [None] if [x] does not respect standards. If contents is an UTF-8
+     contents, [word] will surround [x] with double-quote and will escape
+     control characters (see {!escape_string}).
+
+      NOTE: UTF-8 is allowed in e-mails according RFC 6532. *)
+
   val word_exn : string -> word
+  (** Same as {!word} but raises an exception instead to return [None]. *)
```

Comment:
what exception does this raise?

Reply:
[Invalid_argument] (see [here](https://github.com/mirage/mrmime/blob/factorize/lib/mailbox.ml#L208))